### PR TITLE
[ISSUE #1588] Fix Issue 1588 retry doesn't work when sending msg in async mode and the broker is down

### PR DESCRIPTION
**Pull request by Aaron-He**
Fix bug: retry doesn't work when sending msg in async mode and the
broker is down.

## What is the purpose of the change

Fix bug: retry doesn't work when sending msg in async mode and the
broker is down.

## Brief changelog

Reuse the loop retry logic. This make us have two separate retry logic, one is in charge of resending messages when failed in the sending phrase, the other is in charge of resending when failed in the response-process phrase.

## Verifying this change

yes

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by Aaron-He** [comment]
Thanks @xiangwangcheng for good suggestion. For more information, see #1657

**Comment by zongtanghu** [comment]
We can review this pr together. @duhenglucky

File: client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java [hunk 1]
```diff
@@ -559,7 +559,7 @@ private SendResult sendDefaultImpl(
             MessageQueue mq = null;
             Exception exception = null;
             SendResult sendResult = null;
-            int timesTotal = communicationMode == CommunicationMode.SYNC ? 1 + this.defaultMQProducer.getRetryTimesWhenSendFailed() : 1;
```
Comment (duhenglucky) [inline client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java]: (1) Async Retry used another parameter for retry when send failed: retryTimesWhenSendAsyncFailed
(2) CommunicationMode has there value: SYNC, ASYNC, ONEWAY, and send the message will return directly when using the last two values. so this modification seems has no effect.

Comment (xujianhai666) [reply client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java]: @duhenglucky  So this pr has no effect,  maybe we should close pr?

**Pull request closed without merging**